sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -191,21 +197,24 @@ public void run() {
         if (flushRequested.get() != null) {
           flush();
         }
-
-        try {
-          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
-          if (lastElement != null) {
-            batch.add(lastElement.toSpanData());
-          }
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          return;
+        while (!queue.isEmpty() && batch.size() < maxExportBatchSize) {
+          batch.add(queue.poll().toSpanData());
         }
-
         if (batch.size() >= maxExportBatchSize || System.nanoTime() >= nextExportTime) {
           exportCurrentBatch();
           updateNextExportTime();
         }
+        if (queue.isEmpty()) {
+          try {
+            long pollWaitTime = nextExportTime - System.nanoTime();
+            if (pollWaitTime > 0) {
+              signal.poll(pollWaitTime, TimeUnit.NANOSECONDS);
+            }
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return;
+          }
```

Comment:
Does this produce the same result using drain? Then I think it lets us use `queue.size() >= maxExportBatchSize()` in `addSpan` instead of maintaining a counter.

```java
while(queue.size() >= maxExportBatchSize)) {
  batch.clear();  // Can extract method for these three lines.
  queue.drainTo(batch);
  export(batch);
}
if (!queue.isEmpty() && System.nanoTime() >= nextExportTime) {
  batch.clear();
  queue.drainTo(batch);
  export(batch);
  updateNextExportTime();
} else {
  long pollWaitTime = nextExportTime - System.nanoTime();
  if (pollWaitTime > 0) {
    signal.clear();
    // Small race, queue might reach maxExportBatchSize between these two methods and stop increasing.
    // This is rare in practice and safe since after the poll interval completes the batch will be exported. This
    // behavior is not different from a processor that does not receive spans fast enough to reach the batch
    // size, which is basically what is happening during this interval of time.
    signal.poll(pollWaitTime, TimeUnit.NANOSECONDS);
  }
}
```

Reply:
I guess the race would go away anyways if removing `signal.clear()` from that suggestion - the worst is just that it evaluates the `while` and `if` one more time before sleeping.

Reply:
The logic looks really complicated imho. Repeatedly copying queue into the batch is bit confusing as well as the race. Using a counter we are essentially amortizing the cost of signal, it is intuitive to think that a signal every maxExportBatchSize essentially implies the exporter thread when receiving the signal is guaranteed to do the export.

Reply:
In you suggestion, isn't export(batch) already clearing the batch? Also in the if condition `!queue.isEmpty() && System.nanoTime() >= nextExportTime`, it is exporting the whole batch at once instead of in maxExportBatchSize chunks.

Reply:
>  it is intuitive to think that a signal every maxExportBatchSize essentially implies the exporter thread when receiving the signal is guaranteed to do the export.

But can't the signal be missed, since it's already exporting or something like that?

> In you suggestion, isn't export(batch) already clearing the batch? 

Yeah probably, it's somewhat pseudocode

> Also in the if condition !queue.isEmpty() && System.nanoTime() >= nextExportTime, it is exporting the whole batch at once instead of in maxExportBatchSize chunks.

We've already exported maxExportBatchSize chunks. However, every time the interval passed, we are supposed to clear the queue fully, so this is where remaining spans get exported - it won't export more than maxExportBatchSize even if the queue suddenly had many spans added, because `drainTo` won't fill more than `batch.size()` it's ok.

I think the loop, then if pattern here actually captures the complexity more directly. We have two conditions

- Eagerly export as many spans as possible when over maxExportBatchSize
- If interval has passed, make sure to export entire queue

So these are handled with two conditionals any time the thread wakes up, since there are actually two conditions. Does it make sense?

Let's not worry about the race, if we never clear the signal queue we're good and I think that will work well.

Reply:
> But can't the signal be missed, since it's already exporting or something like that?

Not really, if exporter thread is doing IO then on the next iteration of the loop it copies over any span entries to batch and polls the signal queue, then eventually goes to sleep in the next iteration after eagerly copying over span entries. The guarantee the atomic counter provides is that if exporter thread is sleeping, then the next signal would guarantee that batch size would cross over matchExportBatchSize triggering IO.

> because drainTo won't fill more than batch.size() it's ok
oh so your pseudocode would really be queue.drainTo(batch, matchExportBatchSize)?

> However, every time the interval passed, we are supposed to clear the queue fully
My understanding is that current code always exports in chunks of maxExportBatchSize even when the interval is passed. That makes sense given the naming of the config options `maxExportBatchSize`


```
        try {
          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
          ..

        // only one element is added to the batch in the loop, so this condition either means batch size is equal to maxExportBatchSize or  
        if (batch.size() >= maxExportBatchSize || System.nanoTime() >= nextExportTime) {
          exportCurrentBatch();
          updateNextExportTime();
        }
```


In practice queues either run close to full or close to empty. So if queue is running full at sizes >= maxExportBatchSize because exporter is slow to process IO or something, then we are unnecessarily sending signals. This won't be the case when we signal every maxExportBatchSize, it unties how fast or slow exporter really is.

Reply:
The current code is handling the 2 conditions directly you mentioned anyway with the outer loop `while(continueWork)` and inside it is eagerly copying span entries and exporting with the condition `if (batch.size() >= maxExportBatchSize || System.nanoTime() >= nextExportTime)`.

Reply:
> oh so your pseudocode would really be queue.drainTo(batch, matchExportBatchSize)?

Oops - yeah this is what I meant, sorry

> My understanding is that current code always exports in chunks of maxExportBatchSize even when the interval is passed.

Yeah - so the above comment hopefully clarifies, I meant we never drain more than batch size. But to clarify, it's a bit incorrect to say "always exports in chunks of maxExportBatchSize even when the interval is passed" - it could be a chunk of less than maxExportBatchSize too if interval has passed or else it's possible spans are just never exported.

> then we are unnecessarily sending signals.

I guess I'd still err on oversignaling - perhaps the above point that it's possible to send chunks with size less than the batch size clarifies my worry, that this `%` can get out-of-sync with reality when smaller chunks are sent and it becomes unclear what's going on. If we find the oversignalling to be much, we can follow up with a CAS guard too?

> inside it is eagerly copying span entries and exporting with the condition

I guess I'm mostly worried that the check is on the queue being empty `if (queue.isEmpty()) {`

If the queue is never empty, the BSP thread never cooperatively sleeps for other threads. Maybe it's ok but draining seems like it could reduce the overall time the BSP thread is awake.

Reply:
> chunk of less than maxExportBatchSize
True, I didn't clarify the "or less" part. 

> this % can get out-of-sync with reality when smaller chunks are sent and it becomes unclear what's going on
The current code handles sending chunks with size less than  maxExportBatchSize when interval is passed. I am not sure sure what becomes unclear here?

> If the queue is never empty, the BSP thread never cooperatively sleeps for other threads.
I don't think this would be a concern for two reasons. 1. Exporter thread would either do the IO and get scheduled out by the kernel 2. If exporter is fast then queue would be empty for sure.

> Maybe it's ok but draining seems like it could reduce the overall time the BSP thread is awake.
With drain we would need another array of type ArrayList<ReadableSpan> which makes things bit more cumbersome.

It feels like we would need some experiment data to really see if there are any material impact on the state of exporter thread.

Reply:
> The current code handles sending chunks with size less than maxExportBatchSize when interval is passed. I am not sure what becomes unclear here?

If I'm not mistaken, then if a chunk is sent that isn't `maxExportBatchSize` then the `%` becomes out of sync. The intention of this `%` is to send chunks with size `maxExportBatchSize` but if any batch is sent with a different size due to the interval check, then when the signal is sent, we will have less or more than `maxExportBatchSize` in an unclear way. For example if batch size is 5, and 4 spans are sent due to the interval check, then the signal will happen with only 1 span in the queue. I don't think that's our intention with that signal.

Reply:
I updated the PR to just use an atomic boolean and queue size. I think it will make things much more clear though I didn't see any change in benchmark results.

drainTo() holds up queue lock for a long time which negatively impacts the writer threads.

Reply:
I like this latest iteration quite a bit. Unfortunately, it looks like `BatchSpanProcessorTest.forceExport()` is quite flaky with this implementation. It failed about 1/5 times for me locally. I'm not sure what's going on.

Reply:
@jkwatson could you paste the error? I ran the test 10 times and couldn't reproduce locally.

Reply:
It just fails the first assertion with a '0', rather than a '49'.